### ISSUE TITLE
Event-loop gets blocked with high input-rate

### ISSUE THREAD

**Issue by RobertDiebels** [hyperledger/caliper#263]
As a follow-up to another issue I've been looking at how caliper handles the event-loop.
What I found was that the `updateInterval` in `local-client.js` ceases to be called for high input-rates.
## Context

In `local-client.js` each transaction is started by a call to a rounds' `callback.run()` this in turn calls a blockchains' `invokeSmartContract()` method. For Fabric that results in a `e2eutils.invokebycontext()` call.

This function adds a minimum of 2 + N promises to the micro-tasks queue of the event-loop. Where N is the size of the network/channel. Furthermore it adds a promise for each transaction generated. So when generating 256 Tx/s it adds 256 promises to the micro-tasks queue each second.

Looking at `e2eutils.invokebycontext` we see the following being added to the respective event-queues:
```
channel.sendTransactionProposal -> micro-tasks
eventHubs.forEach() -> 1 promise per eventHub -> micro-tasks
eventHubs.forEach() -> 1 setTimeout per eventHub -> timers queue
eventHubs.forEach() -> 1 IO-callback -> IO events-queue
channel.sendTransaction() -> micro-tasks
await Promise.all(eventPromises) -> micro-tasks
```
The micro-tasks queue is read and only after it is emptied are any timers run. This means that it is possible that the setTimeout per eventHub timer is not processed in a timely manner. Possibly causing a block of in queues further down the line.

## Expected Behavior

I expect caliper not to block the event-loop such that timers are no longer executed.
## Actual Behavior

Caliper blocks the event-loop for high input-rates.
## Possible Fix

A possible fix would be to increase nodes'`UV_THREADPOOL_SIZE` so that file I/O, crypto or other threadpool-dependent tasks are spread across the systems cores. By default nodes' threadpool-size is set to 4.
## Steps to Reproduce

1. Create a series of rounds with high input-rates. (over 140 Tx/s)
2. Run each round for at least 3 minutes.
3. The updatesInterval will no longer be displayed in the console because the event-loop is getting flooded with promises.
4.

## Existing issues

- [ ] [Stack Overflow issues](http://stackoverflow.com/tags/hyperledger-caliper)
- [ ] [GitHub Issues](https://github.com/hyperledger/caliper/issues)
- [ ] [Rocket Chat history](https://chat.hyperledger.org/channel/caliper)

## Context

I was attempting high throughput tests.

## Your Environment

* Version used: N/A
* Environment name and version (e.g. Chrome 39, node.js 5.4):
* Operating System and version (desktop or mobile): Linux
* Link to your project:

**Comment by houqinghui** [comment]
@RobertDiebels this is a good idea, you can fork the caliper  modify it and submit a PR, if you want.

**Comment by RobertDiebels** [comment]
@houqinghui I ended up solving this by increasing the amount of clients. It seems that the loop gets blocked when there are high numbers of transactions (tens of thousands), high input-rates (in the hundreds), high amounts of peers (16+) and low amounts of clients (1-4). You can consider the issue closed afaic. But it might be useful to verify which amounts cause the event loop to start cracking.

I would love to start contributing but it might take a while before I can do so.

**Comment by aklenik** [comment]
@RobertDiebels Yes, this is an inherent issue of NodeJs's programming model. Since it's basically single threaded (so async doesn't mean parallel), there's an inherent limit on the throughput. Imagine the following:
Putting the number of promises and IO wait aside, there are some pure CPU instructions (basic operations) that take some time. If this time amounts to a total of 10ms for each transaction, then setting an input rate above 100TPS will mean that 1s pure CPU time is burned for that 100 Tx. And at this point, the rate controller won't get the program control in time to be accurate. So basically, to maintain a 100TPS load, your program needs to reach the rate controller code every 10ms.

(Fun fact: Fabric 0.x had the same issue with sequential Tx execution :) )

This is mitigated by distributing the load among more clients, so 1 client has "more time" for Tx processing. Usually, I draw the limit around 50, max 70 TPS per client, but this also depends on your hardware.

So to sum it up:
1. Generating high rate workloads is not trivial, and therefore distributed among multiple clients.
2. After a certain rate, you will even reach the limits of a single machine, in which case you distribute the load among multiple clients of multiple machines. This is the long-term plan for Caliper, and we'll get there eventually :)
